spec/datadog/profiling/collectors/code_provenance_spec.rb

```diff
@@ -187,5 +188,76 @@
     it 'renders the list of loaded libraries using the expected schema' do
       JSON::Validator.validate!(code_provenance_schema, code_provenance.generate_json)
     end
+
+    # In PROF-9821 we run into an issue where some versions of OJ + activesupport + monkey patching the JSON gem
+    # would result in our Library instance being encoded instance-field-by-instance-field instead of by calling #to_json.
+    #
+    # This would obviously result in broken code provenance files. To fix this, we've adjusted the class to make sure
+    # that if you serialize it field-by-field, you still get a correct result.
+    #
+    # Reproducing this exact issue in CI is really annoying -- because it would be one more set of appraisails we'd run
+    # just to reproduce it and test.
+    #
+    # So instead in this test we use YAML as an example of an encoder that doesn't use #to_json, and does it
+    # field-by-field. Thus if the Library class doesn't match exactly what we want in the output, this test will fail.
+    #
+    # In case you want to reproduce the exact JSON issue, here's a reproducer:
+    # ````ruby
+    # require 'bundler/inline'
+    #
+    # gemfile do
+    #   source 'https://rubygems.org'
+    #   gem 'activesupport', '= 5.0.7.2'
+    #   gem 'oj', '= 2.18.5'
+    # end
+    #
+    # require 'json'
+    #
+    # class Example
+    #   def initialize = @hello = 1
+    #   def to_json(arg = nil) = {world: 2}.to_json(arg)
+    # end
+    #
+    # example = Example.new
+    # puts JSON.fast_generate(example)
+    #
+    # require 'oj'
+    # require 'active_support/core_ext/object/json'
+    # Oj.mimic_JSON()
+    #
+    # puts JSON.fast_generate(example)
+    # ```
+    #
+    # Incorrect output:
+    # {"world":2}
+    # {"hello":1}
+    #
+    describe 'when JSON encoder is broken and skips #to_json' do
+      let(:library_class_without_to_json) do
+        Class.new(Datadog::Profiling::Collectors::CodeProvenance::Library) do
+          undef to_json
+        end
+      end
+
+      it 'is still able to correctly encode a library instance' do
+        instance = library_class_without_to_json.new(
+          name: 'datadog',
+          kind: 'library',
+          version: '1.2.3',
+          path: '/example/path/to/datadog/gem',
+        )
+
+        serialized_without_to_json = YAML.dump(instance)
```

Comment:
Is the reasoning here that YAML serialization is performed field by field, same as some JSON libraries would serialize, or that libraries like `oj` actually use YAML serialization logic? I am guessing it's the former and if so, a comment to this effect in the code would be helpful.

Reply:
Yes! I've added a comment above the spec stating this:

```
    # So instead in this test we use YAML as an example of an encoder that doesn't use #to_json, and does it
    # field-by-field. Thus if the Library class doesn't match exactly what we want in the output, this test will fail.
```